### PR TITLE
[BAHIR-128] Improve sql-cloudant _changes receiver and fix failing test

### PR DESCRIPTION
**Pull request by emlaver**
_What_
Improve stability of _changes receiver.
Fix test that's failing sporadically in sql-cloudant's `CloudantChangesDFSuite`.

_How_
- Improve performance and decrease testing time by setting batch size to 8 seconds and using `seq_interval` _changes feed option.
- Use getResource to load json files path
- Added Mike Rhodes's `ChangesRowScanner` for reading each _changes line and transforming to GSON's JSON object
- Added Mike Rhodes's `ChangesRow` representing a row in the changes feed

See [BAHIR-128](https://issues.apache.org/jira/browse/BAHIR-128)

### CODE REVIEW

**Comment by ApacheBahir** [comment]
Can one of the admins verify this patch?

File: pom.xml
Comment (ckadner) [inline pom.xml]: this change should not be necessary. the `.idea/` folder should only get created at the project root level. any nested files and folders are already covered. maybe a mishap when setting up IntelliJ?

**Comment by ckadner** [comment]
ok to test

**Comment by ApacheBahir** [comment]
Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/111/

File: pom.xml
Comment (emlaver) [reply pom.xml]: Yea, I think it's extra files in my IntelliJ setup.  I'll rebased my changes.

**Comment by ApacheBahir** [comment]
Build successful

**Comment by ApacheBahir** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/112/

**Comment by ApacheBahir** [comment]
Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:red_circle: Build failed: http://169.45.79.58:8080/job/bahir_spark_pr_builder/113/

**Comment by ApacheBahir** [comment]
Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:red_circle: Build failed: http://169.45.79.58:8080/job/bahir_spark_pr_builder/114/

**Comment by ApacheBahir** [comment]
Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:red_circle: Build failed: http://169.45.79.58:8080/job/bahir_spark_pr_builder/115/

**Comment by ApacheBahir** [comment]
Build triggered. sha1 is merged.

**Comment by ApacheBahir** [comment]
Build started sha1 is merged.

**Comment by ApacheBahir** [comment]
Build triggered. sha1 is merged.

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/116/

**Comment by ApacheBahir** [comment]
Build started sha1 is merged.

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/117/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/118/

**Comment by ckadner** [comment]
retest this please

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java
Comment (ckadner) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java]: @emlaver -- is this an outdated copyright statement? If it is still valid you may need to check with the author and/or IBM if you can contribute this code (or variations of it) to open-source.

I am surprised the RAT check did not catch this.

CC @lresende

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRowScanner.java
Comment (ckadner) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRowScanner.java]: same [comment](https://github.com/apache/bahir/pull/57/commits/5e554103bee8162b85948e219dd4b7fdd7707a30#r155934979) as above

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/119/

**Comment by ckadner** [comment]
I just enabled the RAT check for our Jenkins PR builder.

restest this please

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/120/

**Comment by ckadner** [comment]
Very odd that the RAT license checks succeeded. Apparently it is our `scalastyle` checks which complain about any non-conformant license headers, but those don't cover `*.java` files, so I need to update our `checkstyle` rules for Java sources. But I am still puzzled about RAT letting this pass (@lresende)

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java
Comment (ckadner) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java]: @emlaver -- Java sources should reside under `src/main/java` not `src/main/scala` unless you can convert this code to Scala (preferably)

**Comment by lresende** [comment]
@emlaver Could you please rebase to latest master.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java]: @ckadner It's valid and the author is OK with this existing in the open-source.  From my understanding, we're (Cloudant) allowed to retain the copyright on code we open source, but the license effectively grants free use.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/common/ChangesRow.java]: Fixed in 5da88e8.

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/122/

**Comment by ckadner** [comment]
retest this please

**Comment by ApacheBahir** [comment]
:red_circle: Build failed, see build log for details

**Comment by ApacheBahir** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/123/

**Comment by ckadner** [comment]
@emlaver -- I may have influenced the test execution by running the `mvn test slq-cloudant -q` locally while the Jenkins PR builder was running the same tests (I was using the same Cloudant account as Jenkins). I just restarted the build.

**Comment by ApacheBahir** [comment]
:red_circle: Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:red_circle: Build failed: http://169.45.79.58:8080/job/bahir_spark_pr_builder/124/

**Comment by ckadner** [comment]
retest this please

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/125/

**Comment by ckadner** [comment]
one more for good emasure

retest this please

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/126/

**Comment by ckadner** [comment]
uh, maybe one more, just to be sure :-)

retest this please

**Comment by ApacheBahir** [comment]
:red_circle: Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:red_circle: Build failed: http://169.45.79.58:8080/job/bahir_spark_pr_builder/127/

**Comment by ApacheBahir** [comment]
Uh oh, @emlaver unless someone else was running the same tests concurrently to the last Jenkins test run, there may still work to be done.

retest this please

**Comment by ckadner** [comment]
Uh oh, @emlaver unless someone else was running the same tests concurrently to the last Jenkins test run, there may still be some work to be done.

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/128/

**Comment by ckadner** [comment]
retest this please

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/129/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/130/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/131/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/132/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/133/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/134/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/135/

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/136/

**Comment by ApacheBahir** [comment]
:red_circle: Build failed, see build log for details

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful*: http://169.45.79.58:8080/job/bahir_spark_pr_builder/137/

(*) Test failure in Spark SQL Streaming Akka unrelated

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful: http://169.45.79.58:8080/job/bahir_spark_pr_builder/138/

**Comment by ckadner** [comment]
After more than 10 successful builds I'd say that the flaky `sql-cloudant` tests are more stable now than they have ever been. Good work @emlaver. Once you have confirmed that there is no conflict with the copyright for the new code I am happy to merge this.

**Comment by emlaver** [comment]
@ckadner I've confirmed and there is no conflict.  Let's hold off on merging.  I still have some work I'll commit to this for improved stability.  I'll also need to run my usual larger load tests (>10GB).  And, Rich Ellis on my team will also be reviewing this work.

**Comment by ckadner** [comment]
@emlaver since you are still working on this code, do you think you could take care of those deprecation warnings this time around?  :-)

**Comment by emlaver** [comment]
@ckadner  PR for deprecation warnings: https://github.com/apache/bahir/pull/59.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala
Comment (ricellis) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala]: WDYT about making the `seq_interval` the `bulkSize` instead of hard-coding?

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala
Comment (ricellis) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala]: delete?

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala
Comment (ricellis) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala:128]: I'm amazed this parameter was hard-coded as it seems to be a fairly critical tuning parameter for streaming. I guess this is one for another PR.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala
Comment (ricellis) [inline sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala:40]: I'm a bit confused about this change. Since Spark Streaming is the basis for "real-time" or "continuous applications" doesn't this need to keep listening to the changes feed to wait for more changes?

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/DefaultSource.scala:128]: Yes, this was not visible as part of the initial internal `_changes` work.  I will create a new JIRA item and PR for this.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala:40]: For our internal implementation, we (myself and Mayya) wanted the user to have a snapshot of data to load into Spark.  For that to be possible, we decided to use `continuous` style feed with a doc limit.  With the new _changes implementation from Mike's project, the `normal` feed is stable and works as expected.  I've also lowered the amount of requests/load time by removing the HTTP request for the doc limit since it's not needed with `normal` style _changes feed.
To work with data in "real-time", you can use `CloudantReciever` which creates an eternal changes feed within the Spark Streaming context.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/internal/ChangesReceiver.scala]: Remove in f758996.

File: sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala
Comment (emlaver) [reply sql-cloudant/src/main/scala/org/apache/bahir/cloudant/CloudantChangesConfig.scala]: Yes, I think that's a good idea.  Fixed in f758996.

**Comment by ApacheBahir** [comment]
:white_check_mark: Build successful

**Comment by ApacheBahir** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://169.45.79.58:8080/job/bahir_spark_pr_builder/142/

**Review by ricellis — APPROVED** [review]
+1

**Comment by lresende** [comment]
LGTM Merging if there are no other comments

**Comment by mikerhodes** [comment]
I'm +1 on this PR including the code I originally wrote. Pleased to see it get use in the wild :)

**Pull request closed without merging**
